### PR TITLE
Lvnod 47 auth

### PR DESCRIPTION
**Pull request by AlbaMontis**
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/browse/LVNOD-46)
* [JIRA subtask](https://ssu-jira.softserveinc.com/browse/LVNOD-56)
* [JIRA subtask](https://ssu-jira.softserveinc.com/browse/LVNOD-55)

## Code reviewers

- [ ] @yarvanya
- [ ] @avoytovich
- [ ] @yuliannakorol

### Second Level Review

- [ ] @lhalam
- [ ] @4asof4ik

### CODE REVIEW

**Review by avoytovich — DISMISSED** [review]

**Review by yuliannakorol — DISMISSED** [review]

File: server/app.js [hunk 1]
```diff
@@ -1,11 +1,29 @@
...
 app.use(bodyParser.json());
 app.use(bodyParser.urlencoded({extended: false}));
+app.use(function(req, res, next) {
+  res.header('Access-Control-Allow-Origin', '*');
+  res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
+  next();
+});
+
+app.use(session({
+  secret: secret.key,
```
Comment (4asof4ik) [inline server/app.js]: We don't  need save token in our server side

File: server/controllers/auth.controller.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        const token = req.session.token = jwt.sign({
+          email: user.email,
+          id: user.id
+        }, secret.key);
+        res.json({'token': token});
+      }
+    })
+    .catch(error => res.status(401).send(error));
+  }
+};
```
Comment (4asof4ik) [inline server/controllers/auth.controller.js:22]: Add logout

File: server/routes/index.js [hunk 3]
```diff
@@ -11,4 +12,6 @@ module.exports = app => {
 
   app.post('/api/users', userController.create);
   app.delete('/api/user/:id', userController.destroy);
+
+  app.post('/api/auth/login', authController.login);
```
Comment (4asof4ik) [inline server/routes/index.js:16]: Logout

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: package.json [hunk 4]
```diff
@@ -31,7 +33,7 @@
   "scripts": {
     "build:prod": "NODE_ENV=production webpack -p --config ./webpack/webpack.config.prod.js",
     "start:dev": "webpack-dev-server --hot --inline --config ./webpack/webpack.config.dev.js",
-    "start:prod": "npm run build:prod && node server.js"
+    "start:prod": "npm run build:prod && nodemon server.js"
```
Comment (4asof4ik) [inline package.json]: production?  maybe develop?

File: config/jwt.secretkey.json [hunk 5]
```diff
@@ -0,0 +1,3 @@
+{
+  "key": "damisuperpupermegasicretkey"
```
Comment (4asof4ik) [inline config/jwt.secretkey.json]: this file should be in .gitignore and created manually in our AWS

**Review by 4asof4ik — CHANGES_REQUESTED** [review]

File: server/app.js [hunk 6]
```diff
@@ -1,11 +1,41 @@
...
+          message: 'Failed to authenticate token.'
+        });
+      } else {
+        req.decoded = decoded;
+        next();
+      }
+    });
+
+  } else {
+    return res.status(403).send({
```
Comment (omaik) [inline server/app.js]: 401 is used for unauthorized access.

File: server/controllers/auth.controller.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+module.exports = {
+  login(req, res) {
+    User.findOne({where: {email: req.body.email}}).then(user => {
+      if (!user || !passwordHash.verify(req.body.password, user.password)) {
+        res.status(401).send('Email or password is not valid');
+      } else {
+        const token = jwt.sign({
+          email: user.email,
+          id: user.id
```
Comment (omaik) [inline server/controllers/auth.controller.js]: this token would never expire, which is pretty bad.

Comment (omaik) [reply server/controllers/auth.controller.js]: Also email seems redundant in the sign, id should be enough.

**Review by 4asof4ik — APPROVED** [review]

**Review by denisbodnar — APPROVED** [review]

**Review by avoytovich — APPROVED** [review]

**Pull request merged**